ibis/expr/analysis.py

```diff
@@ -338,48 +338,24 @@ def propagate_down_window(func: ops.Value, frame: ops.WindowFrame):
     return type(func)(*clean_args)
 
 
-# TODO(kszucs): rewrite to receive and return an ops.Node
-def windowize_function(expr, frame):
-    assert isinstance(expr, ir.Expr), type(expr)
-    assert isinstance(frame, ops.WindowFrame)
-
-    def _windowize(op, frame):
-        if isinstance(op, ops.WindowFunction):
-            walked_child = _walk(op.func, frame)
-            walked = walked_child.to_expr().over(op.frame).op()
-        elif isinstance(op, ops.Value):
-            walked = _walk(op, frame)
-        else:
-            walked = op
-
-        if isinstance(walked, (ops.Analytic, ops.Reduction)):
-            return op.to_expr().over(frame).op()
-        elif isinstance(walked, ops.WindowFunction):
-            if frame is not None:
-                frame = walked.frame.copy(
-                    group_by=frame.group_by + walked.frame.group_by,
-                    order_by=frame.order_by + walked.frame.order_by,
-                )
-                return walked.to_expr().over(frame).op()
-            else:
-                return walked
-        else:
-            return walked
-
-    def _walk(op, frame):
-        # TODO(kszucs): rewrite to use the substitute utility
-        windowed_args = []
-        for arg in op.args:
-            if isinstance(arg, ops.Value):
-                arg = _windowize(arg, frame)
-            elif isinstance(arg, tuple):
-                arg = tuple(_windowize(x, frame) for x in arg)
+def windowize_function(expr, default_frame):
+    func = var("func")
+    frame = var("frame")
 
-            windowed_args.append(arg)
```

Comment:
Huzzah! 🎉 

This is waaaaayyyy better.

Reply:
I'm still not satisfied with it. The issue is that we push things through Projector several times and it makes the logic hardly trackable. For example the window merging only happens from the GroupedTable projections, but I should either call `wrap_analytic` rule before passing to Projector with the grouped frame or keep it as is and pass the desired frame as an argument.

Anyway, once we clean up the projector we may be able to further simplify this.